src/databricks/labs/dqx/pii/pii_detection_funcs.py

```diff
@@ -0,0 +1,153 @@
+    """
+    warnings.warn(
+        "PII detection uses user-defined functions which may degrade performance. "
+        "Sample or limit large datasets when running PII detection.",
+    )
+
+    if threshold < 0.0 or threshold > 1.0:
+        raise ValueError(f"Provided threshold {threshold} must be between 0.0 and 1.0")
+
+    if not nlp_engine_config:
+        nlp_engine_config = _default_nlp_engine_config
+
+    if not isinstance(nlp_engine_config, dict | NLPEngineConfig):
+        raise ValueError(f"Invalid type provided for 'nlp_engine_config': {type(nlp_engine_config)}")
+
+    _validate_environment()
+    config_dict = nlp_engine_config if isinstance(nlp_engine_config, dict) else nlp_engine_config.value
+    entity_detection_udf = _build_detection_udf(config_dict, language, threshold, entities)
+    col_str_norm, _, col_expr = _get_normalized_column_and_expr(column)
+    entity_info = entity_detection_udf(col_expr)
+    condition = entity_info.isNotNull()
+    message = concat_ws(" ", lit(f"Column '{col_str_norm}' contains PII:"), entity_info)
+
+    return make_condition(condition=condition, message=message, alias=f"{col_str_norm}_contains_pii")
+
+
+def _validate_environment() -> None:
+    """
+    Validates that the environment can run PII detection checks which use Python dependencies.
+
+    As of Databricks Connect 17.1, strict limits are imposed on the size of dependencies for
+    user-defined functions. UDFs will fail with out-of-memory errors if these limits are exceeded.
+
+    Because of this limitation, we limit use of PII detection checks to local Spark or a Databricks
+    workspace. Databricks Connect uses a `pyspark.sql.connect.session.SparkSession` with an external
+    host (e.g. 'https://hostname.cloud.databricks.com'). To raise a clear error message, we check
+    the session and intentionally fail if `does_not_contain_pii` is called using Databricks Connect.
+    """
+    connect_session_pattern = re.compile(r"127.0.0.1|.*grpc.sock")
+    session = pyspark.sql.SparkSession.builder.getOrCreate()
+    if isinstance(session, pyspark.sql.connect.session.SparkSession) and not connect_session_pattern.search(
+        session.client.host
+    ):
+        raise ImportError("'does_not_contain_pii' is not supported when running checks with Databricks Connect")
+
+
+def _build_detection_udf(
+    nlp_engine_config: dict, language: str, threshold: float, entities: list[str] | None
+) -> Callable:
+    """
+    Builds a UDF with the provided threshold, entities, language, and analyzer.
+
+    :param nlp_engine_config: Dictionary configuring the NLP engine used for PII detection
+    :param language: Language of the text
+    :param threshold: Confidence threshold for named entity detection (0.0 to 1.0)
+    :param entities: List of entities to detect
+    :return: PySpark UDF which can be called to detect PII with the given configuration
+    """
+
+    @pandas_udf("string", PandasUDFType.SCALAR)
```

Comment:
This way of defining pandas udfs will be depreciated in the future. Since Spark 3.0+, the preferred way is to use type hints and omit PandasUDFType.SCALAR entirely. 

Correct way should be:
```
@pandas_udf("string")
def handler(batch: pd.Series) -> pd.Series:
```